### DISCUSSION TITLE
Help use Psalm Phar as LSP in Neovim with no plugins

### DISCUSSION THREAD

**Discussion by still-dreaming-1** [vimeo/psalm#9189] [Q&A]
Theoretically, you don't actually need to use any plugins to use an LSP in Neovim as Neovim has LSP support built in. Checkout `:help LSP`. I'm trying to figure out how to do this, and there seems to be a lack of information about the LSP within the Psalm documentation. It does say how to use it with a few different plugins, but it doesn't really explain what it is showing, and I find it odd that these settings are supposed to work. Specifically the references to "vendor/bin/psalm-language-server" are confusing me. Where is this mysterious psalm-language-server? I don't see it in my project /vendor/bin/, am I supposed to? Perhaps this is because I add Psalm to the project as a phar?

**Comment by weirdan** [comment]
`vendor/bin/psalm-language-server` is available when you install Psalm via composer (`vimeo/psalm`, not `psalm/phar`).

With `psalm.phar` you can run `psalm.phar --language-server` instead.

As for `nvim` config, here's what I use:

```lua
local lsputil = require 'lspconfig/util'
local lspconfig = require 'lspconfig'

lspconfig.psalm.setup{
  cmd = { "vendor/bin/psalm-language-server", "--verbose" },
  filetypes = { "php" },
  root_dir = lsputil.root_pattern("psalm.xml", "psalm.xml.dist"),
  on_attach = on_lsp_attach
}
```
where `on_lsp_attach` is a function that just sets up a couple of keyboard mappings.

The setup above assumes you have https://github.com/neovim/nvim-lspconfig installed.

**Reply by still-dreaming-1** [reply]
I'm not sure what I'm expecting out of you with such a long reply from me. I guess I'm just sharing/ranting/getting stuff of my chest. I tend to become very wordy when I get a keyboard in front of me.

Thanks for the response. I have gone back and forth with many different approaches to getting an LSP to work well in Neovim for many years now in conjunction with Psalm, and I have never actually gotten one working well for me. But I am finally making some progress with my config and getting my knowledge and config up to date with modern Neovim configs and tools, which will make things easier for me when referencing documentation. I am finally using lua for a decent portion of my config and switched over to packer.

It seems like many things work differently with the phar and this creates a barrier for tooling integration for those that use the phar since it won't work out of the box with many integrations unless they specifically make it integrate with the phar as well. Perhaps it would be better if the way you use the language server from Psalm installed the normal way and the way you use it from the phar were the same. With the normal installation Psalm provides a separate location to run it. Perhaps either there could also be a separate phar that runs the language server in that way, or the separate location could be removed, and both install types rely on a terminal parameter.

I have gone back and forth on using a phar and not using one. At first I had no interest in using it, but then I started having dependency clashes within my project. The phar helped with that, but then using the phar itself became problematic, so I switched to just installing Psalm the normal way but in a separate location. But that is also cumbersome to manage, so I eventually switched back to using the phar. Every once in a while I try out going back to just installing Psalm the normal way in our project and notice there is no longer a dependency clash. I recently removed some old dependencies from our project that will never come back again, and for now we don't need the phar. Hopefully it will stick this time.

I have tried using coc.nvim to help me manage my language servers, but I could not get that to work with Psalm at all. It just doesn't even seem to notice that Psalm exists no matter how perfectly I configure it or how standard I get my configuration in harmony with all provided instructions and documentation.  Now I'm not really even liking the idea of using coc.nvim because it is not really the Neovim way. It relies on node.js and is its own LSP client implementation and is even trying to support other VS Code plugins.

I also tried using lsp-zero.nvim. This basically worked out of the box with Psalm, but it doesn't work the way I want it to. When I try to use autocompletion it is offering tons of irrelevant suggestions and claiming they all are coming from the LSP, which can't possibly be true as I know Psalm is not suggesting all of those things, and that is the only php language server I have installed. In VS Code, if I disable some of the built in php support, and install the psalm LSP plugin, everything works really well, at least much better than anything else I have tried. But I just have too much invested into Vim/Neovim at this point, with a config file that is over a thousand lines long and multiple plugins I have developed and used regularly to make my workflow more efficient. I could potentially learn to use VS Code with a Neovim backend using that plugin someone made for this purpose. I tried it briefly and I couldn't even get it to find my Neovim, but I haven't tried it for a long time now. I'm not thrilled about the idea of figuring out how to get Neovim and VS Code to play nice together and create my ideal editing environment. But there are some nice things VS Code brings to the table like the Psalm LSP works how I want, copy and paste uses my local machine's clipboard even if I'm using that remote tool where VS Code runs both on a server and my machine for the UI to be local.

So now I'm torn between further configuring lsp-zero or just bearing down and figuring out how to integrate all this stuff myself so it's not doing a bunch of strange things I never wanted in ways I don't like, like an IDE would. lsp-zero is like 20 different plugins and an entire extension ecosystem just to help you install and configure LSPs using the built in LSP. In a way it is really nice because when I open a file type in Neovim and I don't have an LSP server installed for it, it offers me a list of ones I can install. The fact that it relies on so many different things just to help you configure the built in LSP makes me simultaneously scared of learning to configure the built in LSP myself and scared of learning how to configure all these different plugins.
